components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt

```diff
@@ -312,47 +322,58 @@ class PromptFeature private constructor(
+                    is MenuChoice -> it.onConfirm(value as Choice)
+                    is PromptRequest.Popup -> it.onAllow()
+                    is MultipleChoice -> it.onConfirm(value as Array<Choice>)
+
+                    is Authentication -> {
+                        val (user, password) = value as Pair<String, String>
+                        it.onConfirm(user, password)
+                    }
 
-                is Share -> it.onSuccess()
+                    is TextPrompt -> {
+                        val (shouldNotShowMoreDialogs, text) = value as Pair<Boolean, String>
 
-                is LoginPrompt -> it.onConfirm(value as Login)
+                        promptAbuserDetector.userWantsMoreDialogs(!shouldNotShowMoreDialogs)
+                        it.onConfirm(!shouldNotShowMoreDialogs, text)
+                    }
 
-                is PromptRequest.Confirm -> {
-                    val (isCheckBoxChecked, buttonType) = value as Pair<Boolean, MultiButtonDialogFragment.ButtonType>
-                    promptAbuserDetector.userWantsMoreDialogs(!isCheckBoxChecked)
-                    when (buttonType) {
-                        MultiButtonDialogFragment.ButtonType.POSITIVE ->
-                            it.onConfirmPositiveButton(!isCheckBoxChecked)
-                        MultiButtonDialogFragment.ButtonType.NEGATIVE ->
-                            it.onConfirmNegativeButton(!isCheckBoxChecked)
-                        MultiButtonDialogFragment.ButtonType.NEUTRAL ->
-                            it.onConfirmNeutralButton(!isCheckBoxChecked)
+                    is Share -> it.onSuccess()
+
+                    is LoginPrompt -> it.onConfirm(value as Login)
+
+                    is PromptRequest.Confirm -> {
+                        val (isCheckBoxChecked, buttonType) =
+                            value as Pair<Boolean, MultiButtonDialogFragment.ButtonType>
+                        promptAbuserDetector.userWantsMoreDialogs(!isCheckBoxChecked)
+                        when (buttonType) {
+                            MultiButtonDialogFragment.ButtonType.POSITIVE ->
+                                it.onConfirmPositiveButton(!isCheckBoxChecked)
+                            MultiButtonDialogFragment.ButtonType.NEGATIVE ->
+                                it.onConfirmNegativeButton(!isCheckBoxChecked)
+                            MultiButtonDialogFragment.ButtonType.NEUTRAL ->
+                                it.onConfirmNeutralButton(!isCheckBoxChecked)
+                        }
                     }
                 }
+            } catch (e: ClassCastException) {
+                crashReporting?.recordCrashBreadcrumb(
+                    Breadcrumb("PromptFeature onConsume cast failed",
+                        hashMapOf("class name" to " ${it.javaClass}"), "crash",
+                        Breadcrumb.Level.DEBUG, Breadcrumb.Type.NAVIGATION)
+                )
+
+                crashReporting?.submitCaughtException(e)
```

Comment:
I don't really like the idea of catching `ClassCastException` here every time since it will just hide errors while developing. Maybe we could just consider adding regular breadcrumbs to the feature (showing prompt, confirming, cancelling) that are helpful for understanding all kinds of errors?

Reply:
Sure, I'll add breadcrumb and throw the exception after.

Reply:
Oh, I was thinking about just generic breadcrumbs independent of the exception here.

But talking about the exception. Maybe you do not want a breadcrumb and instead just rethrow a new exception with a better message (and the original one as cause)?

Reply:
I am worried this is called too many times to add a generic breadcrumb without an exception.  Sure, I could modify the message.  But wouldn't that be similar to adding a breadcrumb since we will be try catching the exception and rethrowing?